### ISSUE TITLE
"chezmoi import" leading to a "gzip: invalid header" error

### ISSUE THREAD

**Issue by LittleBigDev** [twpayne/chezmoi#1318]
## What exactly are you trying to do?
I have a (private) gitlab project where I can download a tarball to be imported : 
```bash
curl -s -L -o ${TMPDIR:-/tmp}/my_project-master.tar.gz https://self.hosted.gitlab/my_namespace/my_project/-/archive/master/my_project-master.tar.gz
chezmoi import --strip-components 1 --destination ${HOME}/my_project ${TMPDIR:-/tmp}/my_project-master.tar.gz
```

## What have you tried so far?

I have tested commands individually :
```bash
curl -s -L -o ${TMPDIR:-/tmp}/my_project-master.tar.gz https://self.hosted.gitlab/my_namespace/my_project/-/archive/master/my_project-master.tar.gz
```
=> the tarball exists in my tmp dir.

```bash
tar zxvf ${TMPDIR:-/tmp}/my_project-master.tar.gz
```
=> tarball is successfully extracted

But still the `chezmoi import` command fails :
```bash
$ chezmoi import --strip-components 1 --destination ${HOME}/my_project ${TMPDIR:-/tmp}/my_project-master.tar.gz        
chezmoi: gzip: invalid header
$
```
```bash
$ chezmoi import --destination ${HOME}/my_project ${TMPDIR:-/tmp}/my_project-master.tar.gz        
chezmoi: gzip: invalid header
$
```
=> always the invalid header error.

## Where else have you checked for solutions?

* [x] I have read [chezmoi's how-to guide](https://github.com/twpayne/chezmoi/blob/master/docs/HOWTO.md), and not found the answer.
* [x] I have read [chezmoi's FAQ](https://github.com/twpayne/chezmoi/blob/master/docs/FAQ.md), and not found the answer.
* [x] I have searched [chezmoi's reference guide](https://github.com/twpayne/chezmoi/blob/master/docs/REFERENCE.md), and not found the answer.
* [x] Other, please give details. (google, stackoverflow etc... No interresting result)

## Output of any commands you've tried with `--verbose` flag

```bash
$ chezmoi --verbose import --destination ${HOME}/my_project ${TMPDIR:-/tmp}/my_project-master.tar.gz        
chezmoi: gzip: invalid header
$
```
=> no more info

## Output of `chezmoi doctor`

(I edited my username)
```console
$ chezmoi doctor
RESULT   CHECK                MESSAGE
ok       version              v2.1.2, commit a3fb755a88315bbdecedba3d849324752989400d, built at 2021-07-14T20:07:22Z, built by goreleaser
ok       os-arch              linux/amd64 (Ubuntu/20.04.2 LTS (Focal Fossa))
ok       config-file          /home/<myuser>/.config/chezmoi/chezmoi.toml is a file
ok       source-dir           /home/<myuser>/.local/share/chezmoi is a directory
ok       suspicious-entries   no suspicious entries found in /home/<myuser>/.local/share/chezmoi
ok       dest-dir             /home/<myuser> is a directory
ok       shell                found /usr/bin/zsh
ok       editor               found /usr/bin/vi
ok       git-cli              found /usr/bin/git, version 2.25.1
ok       merge-cli            found /usr/bin/vimdiff
ok       age-cli              age not found in $PATH
ok       gnupg-cli            found /usr/bin/gpg, version 2.2.19
info     1password-cli        op not found in $PATH
info     bitwarden-cli        bw not found in $PATH
info     gopass-cli           gopass not found in $PATH
info     keepassxc-cli        keepassxc-cli not found in $PATH
info     keepassxc-db         not set
ok       lastpass-cli         found /usr/bin/lpass, version 1.3.3
info     pass-cli             pass not found in $PATH
info     vault-cli            vault not found in $PATH
info     secret-cli           not set

```

## Additional context

- I was not able to find the details involved in `chezmoi import` command, so I could not troubleshoot it step by step.
- I supposed that trying to `tar zxvf` my tarball would be enough confirm the file is ok. But maybe this is not enough.
- I was able to use `chezmoi import` with tarballs from public github repositories (robbyrussell/oh-my-zsh and zsh-users/zsh-syntax-highlighting) with no problem.

**Comment by twpayne** [comment]
I suspect that the tarballs generated by gitlab are not compatible with the Go standard library's gzip package. Could you attach an example tarball please?

**Comment by LittleBigDev** [comment]
I created a fake gitlab project (same server, same group) with some files and tried to import the tarball I downloaded from it. No problem importing it.
I signed a confidentiality clause, so I can't send you the tarball. I think this is a project-related issue, maybe some weird files in there...  
Anyway, this is out of your scope. Your import feature works just fine with any other tarball. So I guess it will stay a mystery :)

I was able to work around : an sh script cloning and copying the needed files. Not declarative, I know. But well. It works ¯\_(ツ)_/¯

**Comment by twpayne** [comment]
OK, but I prefer understanding to mysteries. If you find out what the issue was, please do enlighten me!

chezmoi [uses the file extension to determine how to decompress it](https://github.com/twpayne/chezmoi/blob/master/internal/cmd/importcmd.go#L68-L79). You could download the `.tar.gz`, decompress it with `gzip`, and then pass the resulting `.tar` to `chezmoi import`, and it should work.

**Comment by LittleBigDev** [comment]
Ok I dug out a bit more, and it looks like when I download the tarball via my browser (I click the DL link and save the tar.gz on my disk), then chezmoi imports it without any issue. When I download it with curl though, the import fires the gzip error.

First I thought it was about the /tmp/ dir. But I had the same result with my home dir. 

TL;DR : 
- `curl -s -L -o /whatever/path/to/tarballtest-master.tar.gz https://download.url.for/tarballtest-master.tar.gz` => `chezmoi import` leads to "chezmoi: gzip: invalid header"
- Download the same tarball via my browser : chezmoi imports it succesfully.

Tell me if you need anything more from me !

**Comment by LittleBigDev** [comment]
Small update : using HTTPie instead of cURL had the same result (`chezmoi: gzip: invalid header`)
I can't use wget because I get a 401 error (server is reachable only over VPN, with private key authentication and I wasn't able to make wget work like this)

**Comment by twpayne** [comment]
When you download the file with `curl`, what do the following commands print?

    file /whatever/path/to/tarballtest-master.tar.gz

    head -n 5 /whatever/path/to/tarballtest-master.tar.gz

**Comment by LittleBigDev** [comment]
`file` : "ASCII text, with no line terminators"
`head -n 5` : "You need to sign in or sign up before continuing.%"

Looks like I need to have an active logged session before direct downloading the tarball.

Also, looks like I made tests on the wrong file yesterday. Today I tried again to `tar zxvf` the "curl" one, and had an error : 
```
gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

Confirmed with : 
```
$ cat tarballtest-master.tar.gz 
You need to sign in or sign up before continuing.% 
```

**Comment by LittleBigDev** [comment]
Fixed, using an access token.
```
$ curl --header 'PRIVATE-TOKEN: <secret>' -s -L -o ~/tarballtest-master.tar.gz https://path/to/tarballtest/-/archive/master/tarballtest-master.tar.gz
$ tar zxvf ~/tarballtest-master.tar.gz
tarballtest-master/
tarballtest-master/README.md
```
My dawnload & import script works fine with the token.

Problem solved :)
